### PR TITLE
Fix clang-tidy errors in RegexAST.hpp; Add Microsoft GSL as a dependency.

### PR DESCRIPTION
**Pull request by SharafMohamed**
- Add GSL for gsl::at and gsl::owner
- Use [maybe unused]
- Use std::array in place of c-array
- Follow rule of 5
- Did not move RegexAST.tpp into RegexAST.hpp yet to reduce the size of the PR. In following PR will fix RegexAST.tpp, then another PR to combine them.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -106,6 +120,7 @@ target_compile_definitions(log_surgeon
 install(
     TARGETS
     log_surgeon
+    GSL
```
Comment (kirkrodrigues) [inline CMakeLists.txt:123]: Let's continue discussing this offline.

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

File: CMakeLists.txt [hunk 2]
```diff
@@ -72,12 +81,17 @@ set(LCHIP_INSTALL_INCLUDE_DIR ${CMAKE_INSTALL_INCLUDEDIR})
...
+                      PUBLIC
+                      Microsoft.GSL::GSL
+)
 target_include_directories(log_surgeon
     PUBLIC
     $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src>
     $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
     PRIVATE
     ${CMAKE_CURRENT_SOURCE_DIR}/src
+    ${GSL_SOURCE_DIR}/include
```
[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply CMakeLists.txt]: Fixed.

File: src/log_surgeon/finite_automata/RegexAST.hpp [hunk 3]
```diff
@@ -2,35 +2,43 @@
...
 
 template <typename NFAStateType>
 class RegexAST {
 public:
     virtual ~RegexAST() = default;
+    RegexAST() = default;
+    RegexAST(RegexAST const& rhs) = default;
+    auto operator=(RegexAST const& rhs) -> RegexAST& = default;
+    RegexAST(RegexAST&& rhs) noexcept = default;
+    auto operator=(RegexAST&& rhs) noexcept -> RegexAST& = default;
```
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply src/log_surgeon/finite_automata/RegexAST.hpp:27]: Used protected instead

File: src/log_surgeon/finite_automata/RegexAST.hpp [hunk 4]
```diff
@@ -168,30 +180,33 @@ class RegexASTGroup : public RegexAST<NFAStateType> {
...
      * @param is_possible_input
      */
-    auto set_possible_inputs_to_true(bool is_possible_input[]) const -> void override {
+    auto set_possible_inputs_to_true(std::array<bool, cUnicodeMax> is_possible_input
+    ) const -> void override {
         if (!m_negate) {
-            for (Range range : m_ranges) {
-                for (uint32_t i = range.first; i <= range.second; i++) {
-                    is_possible_input[i] = true;
+            for (auto const& [fst, snd] : m_ranges) {
```
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply src/log_surgeon/finite_automata/RegexAST.hpp:195]: Begin + end is fine, this was just CLion auto-formatting it to fst and snd.

**Review by kirkrodrigues — APPROVED** [review]

**Comment by kirkrodrigues** [comment]
For the PR title, how about:

Fix clang-tidy errors in RegexAST.hpp; Add Microsoft GSL as a dependency.

**Pull request merged**